### PR TITLE
[RFC] Move out of CRA and use NextJS

### PR DESCRIPTION
**Pull request by angelocordon**
## What type of PR is this? (check all applicable)

- [x] ♻️ Refactor
- [x] 🎨 Enhancement
- [x] 🚩 Other

## Context

Experimenting with NextJS, as a way for us to structure our front end application a bit more clearly. There are some interesting patterns that I'm currently exploring: 
1/ page structures - which has something that I've been pushing for some time with our past React apps. Page components "shouldn't" live inside a components directory, as it makes it difficult to identify from a Page/Route to a purely UI Component.
https://nextjs.org/docs/basic-features/pages

2/ Asynchronous data fetching as props to a page - https://nextjs.org/docs/basic-features/data-fetching#simple-example

3/ API Routes with REST and GraphQL https://nextjs.org/docs/api-routes/introduction

### Other Benefits
Other more meta benefits I'm looking into are:
1/ From a software development standpoint, it has some pretty well-structured opinions that are digestible. I think these kinds of things are good for greener developers to give them the guard rails for developing full-fledged apps, in comparison to CRA's loose structure (read, can be easy to do bad things in CRA). 

2/ From a career experience perspective, NextJS is a popular enough framework that there is a [job market](https://www.linkedin.com/jobs/search/?currentJobId=2155063490&keywords=next%20js) for it. Gaining experience with a professional software development framework through OSS and a safe environment to learn can help folks in the community.

## Screenshots/Recordings (if there are UI changes)

image

Page components can dynamically define routes and abstract away the use of / need for React Router. Files and folders inside the `pages/` directory become routes themselves, i.e.: /pages/about/index.js => `/about`. This also works with /pages/about.js => `/about`

## Other Things to be Considered

- [ ] It's possible to deploy a NextJS application to Netlify. Need to carve out some time to do this next (🥁). However, I'm unsure of the implications of "serverless" fetching of data (as Netlify does not have a Node server)? I'm not sure it's necessary but tagging @d3vild06 here in case he might have some insights from his POC. 

- [ ] Alternatively, should research if Vercel/Zeit is comparable to what we can get out of Netlify either through an Open Source plan or their free tier. 
https://vercel.com/pricing
https://www.netlify.com/legal/open-source-policy/

## Related Tickets & Documents (Optional)

Handles #76 a bit, but mainly by side effect.

## Added tests?

- [x] 🙋 no, because I need help

## Added to documentation (readme.md or contributing.md)?

- [x] 🙅 no documentation needed as its currently a spike; should we move forward, the README should be updated.

### CODE REVIEW

**Comment by lpatmo** [comment]
Do you it's worth making this a separate repo so that it can have its own deployment URL on Netlify (or Vercel)?

> 2/ Asynchronous data fetching as props to a page - https://nextjs.org/docs/basic-features/data-fetching#simple-example

From what I can tell NextJS is pretty opinionated... we would be using staticProps instead of the React Query library, right?

> 1/ page structures - which has something that I've been pushing for some time with our past React apps.

This was actually something I was hoping to do after the CREATE/Search PRs were merged in too. Don't think it's too difficult to do... just a bunch of import paths changing. :) 

> 3/ API Routes with REST and GraphQL https://nextjs.org/docs/api-routes/introduction

We can use NextJS without GraphQL I know, but are you advocating for exploring GraphQL? If so, I think that also supports creating a new repo for this exploration.

=====
fwiw I've played around with NextJS and I have to say I'm not a huge fan of how opinionated it is, although I'm not against exploring it (preferably in a separate repository) and figuring out if we all like it. The _main_ criteria I think we should consider is whether it's actually easier for new people to pick it up compared to CRA.

**Comment by angelocordon** [comment]
> Do you it's worth making this a separate repo so that it can have its own deployment URL on Netlify (or Vercel)?

Our deployments are based on branches, so we can direct the Netlify or Vercel web hooks to listen to a specific branch. The preview URLs should show what’s on the branch that it’s configured to pull from. That said, I’m not positive why we need to fork to a separate repo but let me know if there are other issues I’m not seeing. 

>  we would be using staticProps instead of the React Query library, right

Correct, although we don’t have to. Still, considering that they are practically the same thing, not sure why we’d need to work against the framework for something that comes out of the box. 

> We can use NextJS without GraphQL I know, but are you advocating for exploring GraphQL?

We currently don’t have an immediate need for GraphQL but I can see it working out for us eventually. Not something I want to do right now but Next would at least set us up to implement GraphQL with it calling to DRF’s REST endpoints and serving graph data to the frontend end. Unsure if Next has its own way of handling that or we’d need Apollo. From my research, I’ve found that since Django isn’t super keen in serving up GraphQL, we can keep it as a REST API, and have a Node-based GraphQL middle layer to glue both stacks. NextJS would come into play here with a more structured set up to handle that for us. 

> If so, I think that also supports creating a new repo for this exploration.

Out of curiosity, what is the goal with creating a new repo? Could branching for the same thing for us? Having it here would help us keep close to the current state of the code base  

> fwiw I've played around with NextJS and I have to say I'm not a huge fan of how opinionated it is,

From your experiences, what did you struggle with the most? What didn’t work well for you?

> The main criteria I think we should consider is whether it's actually easier for new people to pick it up compared to CRA.

I don’t think this would work well for people who have never created a React app before. I think it would work better for folks who want to go deeper with application development but don’t have enough professional experience. 

However, this is a stretch goal and let’s not base our opinions fully on this - I think we should evaluate it from the perspective of a framework that can help our own product development.

**Comment by lpatmo** [comment]
> Our deployments are based on branches, so we can direct the Netlify or Vercel web hooks to listen to a specific branch.
👀 Ohh, let's do that then. I saw that in this PR you wanted to merge to `main`, which is talking to Netlify. If NextJS vs. CRA were in separate branches that would be fine. 

> From your experiences, what did you struggle with the most? What didn’t work well for you?
The part that I disliked was simply https://nextjs.org/docs/basic-features/data-fetching, haha. I thought it was harder to understand compared to React Query (or even data fetching in a simple useEffect). 

> I don’t think this would work well for people who have never created a React app before. 
On this point... I think if we establish the best practices for a React app (which we've started doing but are not quite done with), then folks will likely follow those best practices. To me it's the difference between learning React hooks (where there are a lot of tutorials) _plus_ NextJS (framework docs) vs. simply learning React, and following the patterns in this repo. 

> I think we should evaluate it from the perspective of a framework that can help our own product development.
That's also a good criteria -- i.e. what parts of NextJS are easier to speed building things up with vs. CRA? Example: validation?

**Comment by tgrrr** [comment]
> We can use NextJS without GraphQL I know, but are you advocating for exploring GraphQL?
>> We currently don’t have an immediate need for GraphQL but I can see it working out for us eventually. Not something I want to do right now but Next would at least set us up to implement GraphQL with it calling to DRF’s REST endpoints and serving graph data to the frontend end. Unsure if Next has its own way of handling that or we’d need Apollo. From my research, I’ve found that since Django isn’t super keen in serving up GraphQL, we can keep it as a REST API, and have a Node-based GraphQL middle layer to glue both stacks. NextJS would come into play here with a more structured set up to handle that for us.

- I think we need to be able to work with GraphQL sooner rather than later
- My general approach is to wrap REST API queries in GraphQL queries, which means both work, and we wont end up with a redundant API (in the same way we will between an rest `api/v1` and `api/v2`)

> Our deployments are based on branches, so we can direct the Netlify or Vercel web hooks to listen to a specific branch.
👀 Ohh, let's do that then. I saw that in this PR you wanted to merge to main, which is talking to Netlify. If NextJS vs. CRA were in separate branches that would be fine.

There's nothing stopping us from using Netlify with previews on NextJS, CRA, or a React App with a custom Webpack.

> 1/ page structures - which has something that I've been pushing for some time with our past React apps.

>> This was actually something I was hoping to do after the CREATE/Search PRs were merged in too. Don't think it's too difficult to do... just a bunch of import paths changing. :)

Here here. This is definitely something we need to do.

> we would be using staticProps instead of the React Query library, right

>> Correct, although we don’t have to. Still, considering that they are practically the same thing, not sure why we’d need to work against the framework for something that comes out of the box.

We should consider how each library handle side-effects similar to how `redux-saga` handles it

> since Django isn’t super keen in serving up GraphQL

This hasn't been my experience with GraphQL using [Graphene](https://docs.graphene-python.org/projects/django/en/latest/tutorial-plain/). 

What problems have you had @angelocordon ?

---

New point:

NextJS's baked in SSR is a distinct advantage

**Review by tgrrr — COMMENTED** [review]
This is an initial review of the code style

- Tests
- Detailed user testing to ensure the app works across devices

I can spend some more time on the above in the near future, let me know if it's required

File: src/components/Nav/index.js [hunk 1]
```diff
@@ -77,26 +79,26 @@ export default function Nav() {
...
-          <li>
-            <Link to="/groups">Groups</Link>
+            <Link href="/about">
+              <Button component="a">About</Button>
+            </Link>
           </li>
           <li>
-            <Link to="/projects">Projects</Link>
+            <Link href="/resources">
+              <Button component="a">Resources</Button>
```
[the remark this answers is not in the document]
Comment (angelocordon) [reply src/components/Nav/index.js:100]: Thanks, these are good feedback but not quite important in this PR. For clarity, this PR is only meant to serve as a code discussion to explore if NextJS will be viable for us. This PR is not going to be merged in, but when we do move over, we can address these issues separately 👍🏽

**Comment by angelocordon** [comment]
re: Graphene and the Python/Django + GraphQL Ecosystem

From what I can tell, the Python community isn't so bought in with GraphQL. I've also heard that Graphene isn't very performant with serving graph data in comparison to just serving it up the REST way 🤷🏽‍♂️ 

That said, GraphQL in Next is [a whole other problem](https://github.com/vercel/next.js/discussions/10946), I'd rather kick that can further down the road for a while until we really need to address it. I don't think we have an immediate need for it because the data shapes we're getting back from DRF are super simple and small. 

@lpatmo  regarding criteria - what kind of metrics would be beneficial here to try to measure against? For me, some of the wins are routing (more automated, not needing to deal with react-router). The project structure comes out of the box, but that's not something unique to Next. I actually find Next's data-fetching way more straight forward than trying to follow code that fetches with useEffect and comparable to React-Query. Maybe that's something that we can explore deeper? Unsure if it truly does feel confusing or just not enough readable patterns (NOTE: this isn't implemented in this code, so will experiment with it next).

**Comment by lpatmo** [comment]
> what kind of metrics would be beneficial here to try to measure against?

- how quickly beginner React developers can pick it up
- how fast we can develop with it (e.g. routes, data fetching, form validation)
- job-readiness (e.g. learning NextJS framework vs. learning React Router and using React Query pattern, which are more typically used in React projects that don't use a framework)

> NOTE: this isn't implemented in this code, so will experiment with it next

Cool, curious to hear how you find it! :)

**Comment by tgrrr** [comment]
> From what I can tell, the Python community isn't so bought in with GraphQL. I've also heard that Graphene isn't very performant with serving graph data in comparison to just serving it up the REST way ‍♂️

GraphQL can be less performant on a server as REST, but the total processing time is a fraction of the network call to the server.

So, just say we have to make several REST API calls at the same time to multiple endpoints. Eg. Auth + get a list, then return results based on that list. The last place I did some work for had 80 REST API calls on most user interactions!
We can achieve the same thing with a single API call with Graph.

It's a similar argument to SSR, which I know you're familiar with.

> how fast we can develop with it (e.g. routes, data fetching, form validation)
Where are we doing the form validation? 

> [ ] job-readiness
I'm well versed in WebPack including optimisation, and would be happy to handle this task.
I think we should put a custom build in the ring to be selected from.
I don't know a single job I've worked on that uses CRA or NextJS, it's all bespoke.
This wont have an effect on how quickly beginners pick it up, or how fast we can develop with it.

I know this complicates the decision by choosing between 3 frameworks rather than 2, but it's a large structural change, and I think it should be done right

**Comment by tgrrr** [comment]
Is NextJS happening? @angelocordon @lpatmo I feel issues like https://github.com/codebuddies/frontend/issues/76 are dependent on this decision

**Comment by angelocordon** [comment]
I’m still trying to carve out the time to come back at this. Hopefully this week. 

#76 shouldn’t blocked by this as that can be achieved with creating a pages directory and moving the proper components in it. If any, creating a pages folder would help make the transition better should we decide to go down that path as having a pages directory is a requirement for Next (I think you know this already?). 

What other kind of decisions do you feel are dependent on this issue? Maybe I’m missing something?

**Pull request closed without merging**